### ISSUE TITLE
Missing localStorage crashes application

### ISSUE THREAD

**Issue by lukaszbros** [sahat/satellizer#325]
I have an issue with Satellizer library generating errors while initialising application without localStorage available. I use external tool wkhtmltopdf as a printing mechanism for some of the pages. Unfortunately it does not support local storage. Due to lack of check for localStorage on the Satellizer side whole application is not able to initialise and I got blank page as result. The part that is failing is
```js
.config(['$httpProvider', 'satellizer.config', function($httpProvider, config) {
      $httpProvider.interceptors.push(['$q', function($q) {
        var tokenName = config.tokenPrefix ? config.tokenPrefix + '_' + config.tokenName : config.tokenName;
        return {
          request: function(httpConfig) {
            var token = localStorage.getItem(tokenName);
            if (token && config.httpInterceptor) {
              token = config.authHeader === 'Authorization' ? 'Bearer ' + token : token;
              httpConfig.headers[config.authHeader] = token;
            }
            return httpConfig;
          },
          responseError: function(response) {
            return $q.reject(response);
          }
        };
      }]);
    }]);
```
I do not require Satellizer running for printing purposes but as this is a config section I cannot avoid this code to be executed. I do not even use interceptor in my application and have config.httpInterceptor set to false but token query is done before check.

It would be great if check for config.httpInterceptor would be done earlier or localStorage check would be added.

**Comment by Vad1mo** [comment]
+1

**Comment by wololock** [comment]
+1

I run into the same issue, would be great having it fixed.

**Comment by sahat** [comment]
Before I bump the version can someone review the changes I've pushed to `master`?

I wanted to create a `factory` for handling localstorage methods but then I wouldn't be able to use it inside `.config` block.

So I ended up creating a singleton object called `Storage` outside of Satellizer's module, but inside IIFE. All it does is wrap `getItem`, `setItem` and `removeItem` methods of LS. It also checks if LS is present. If not, returns a fake objects with empty  `getItem`, `setItem` and `removeItem` methods and prints a warning to the console.

![screenshot 2015-03-18 19 14 19](https://cloud.githubusercontent.com/assets/544954/6723323/8488df48-cda3-11e4-9f15-7037cefd8c3d.png)

```js
var Storage = (function() {

    function supportsLocalStorage() {
      try {
        return 'localStorage' in window && window['localStorage'] !== null;
      } catch(e){
        return false;
      }
    }

    function storage() {
      return {
        get: function(key) {
          return localStorage.getItem(key);
        },
        set: function(key, value) {
          return localStorage.setItem(key, value);
        },
        remove: function(key) {
          return localStorage.removeItem(key);
        }
      };
    }

    function fakeStorage() {
      return {
        get: function() {},
        set: function() {},
        remove: function() {}
      };
    }

    if (supportsLocalStorage()) {
      return storage();
    } else {
      console.warn('Warning: Browser Local Storage is disabled or unavailable. Satellizer will not work correctly.');
      return fakeStorage();
    }

  })();
```

**Comment by lukaszbros** [comment]
I've just tested latest Stellizer master against wkhtmltopdf and pdf generation worked correctly. So this solves my issue. Thnx.

**Comment by sahat** [comment]
Thanks, I forgot that you can push a `factory` to the interceptor.

Here is the refactored version. It is now on `master`.

```js
.factory('satellizer.storage', function() {
      if (supportsLocalStorage()) {
        return {
          get: function(key) { return localStorage.getItem(key); },
          set: function(key, value) { return localStorage.setItem(key, value); },
          remove: function(key) { return localStorage.removeItem(key); }
        };
      } else {
        console.warn('Warning: Browser Local Storage is disabled or unavailable. Satellizer will not work correctly.');
        return {
          get: function(key) {},
          set: function(key, value) {},
          remove: function(key) {}
        };
      }

      function supportsLocalStorage() {
        try {
          return 'localStorage' in window && window['localStorage'] !== null;
        } catch(e){
          return false;
        }
      }
    })
    .factory('satellizer.interceptor', [
      '$q',
      'satellizer.config',
      'satellizer.storage',
      function($q, config, storage) {
        var tokenName = config.tokenPrefix ? config.tokenPrefix + '_' + config.tokenName : config.tokenName;
        return {
          request: function(httpConfig) {
            var token = storage.get(tokenName);
            if (token && config.httpInterceptor) {
              token = config.authHeader === 'Authorization' ? 'Bearer ' + token : token;
              httpConfig.headers[config.authHeader] = token;
            }
            return httpConfig;
          },
          responseError: function(response) {
            return $q.reject(response);
          }
        };
      }])
    .config(['$httpProvider', function($httpProvider) {
      $httpProvider.interceptors.push('satellizer.interceptor');
    }]);
```

**Comment by sahat** [comment]
In the future, I may add `storageType` property where you can choose between cookies, local storage and chrome storage (chrome extensions).

**Comment by lukaszbros** [comment]
I've just tested and factory version also works for my purposes. Thnx.

**Comment by sabrehagen** [comment]
Thanks Sahat!

**Comment by sabrehagen** [comment]
I should also alert you to this before this pr is merged: #330

**Comment by lukaszbros** [comment]
Is there a timeline for releasing this fix?
